### PR TITLE
Emit YAML for events cahcemgr report

### PR DESCRIPTION
**Pull request by kinkie**
Also EventScheduler::dump to be stream-based,
and simplify TestEvent::testDump

### CODE REVIEW

**Comment by kinkie** [comment]
Single-worker output in master:
```
Last event to run: MaintainSwapSpace

Operation                   Next Execution  Weight  Callback Valid?
logfileFlush                0.039 sec       1    yes
MaintainSwapSpace           0.039 sec       1    N/A
fqdncache_purgelru          2.429 sec       1    N/A
ipcache_purgelru            7.531 sec       1    N/A
memPoolCleanIdlePools       13.968 sec      1    N/A
statAvgTick                 38.488 sec      1    N/A
peerRefreshDNS              100.777 sec     1    N/A
peerClearRR                 216.965 sec     0    N/A
netdbSaveState              1711.270 sec        1    N/A
```

Output with PR applied:
```
last event to run: MaintainSwapSpace
scheduled events:
  - operation: logfileFlush
    secs to next execution: 0.424
    weight: 1
    callback valid: yes
  - operation: MaintainSwapSpace
    secs to next execution: 0.438
    weight: 1
    callback valid: N/A
  - operation: fqdncache_purgelru
    secs to next execution: 4.427
    weight: 1
    callback valid: N/A
  - operation: memPoolCleanIdlePools
    secs to next execution: 4.427
    weight: 1
    callback valid: N/A
  - operation: ipcache_purgelru
    secs to next execution: 9.428
    weight: 1
    callback valid: N/A
  - operation: statAvgTick
    secs to next execution: 49.413
    weight: 1
    callback valid: N/A
  - operation: peerClearRR
    secs to next execution: 289.391
    weight: 0
    callback valid: N/A
  - operation: peerRefreshDNS
    secs to next execution: 2973.609
    weight: 1
    callback valid: N/A
```

**Comment by kinkie** [comment]
This PR has not adopted spaces() from PR #1735 yet

File: src/event.cc [hunk 1]
```diff
@@ -258,21 +261,26 @@ EventScheduler::clean()
...
+             "    secs to next execution: " << std::setprecision(3) << std::fixed << (e->when? e->when - current_dtime : 0) << '\n' <<
+             "    weight: " << e->weight << '\n' <<
+             "    callback valid: ";
+        if (e->arg && e->cbdata) {
+            if (cbdataReferenceValid(e->arg))
+                yaml << "yes\n";
+            else
+                yaml << "no\n";
+        } else {
+            yaml << "N/A\n";
```
Comment (rousskov) [inline src/event.cc:282]: Just like in other reports you have already converted, please do not report fields that are not applicable to an event. It was difficult or even wrong to skip a field when this was a table. This PR is no longer using a table, removing that obstacle.

**Review by rousskov — CHANGES_REQUESTED** [review]
Please settle spaces() issue _before_ posting any new YAML PRs or requesting additional reviews on the already posted ones. Squid currently lacks the necessary foundation for new YAML PRs, and we have enough information to provide that foundation.

**Comment by kinkie** [comment]
This is ready to review, using a local indent SBuf instead of spaces() while waiting for spaces to land. But apart from that bit of rework, I don't have any outstanding requests for change

File: src/event.cc (hunk 1 above)
Comment (rousskov) [reply src/event.cc]: > for me spaces() can go in as-is. What do I need to do to convince you that it's good enough? 

This PR is not the right place to discuss problems with spaces().

> Right now I have a backlog of completed changes

Please do not interpret my commends below as an implicit agreement with the above "completed" characterization of any pending changes.

Creating more and more non-urgent PRs (or code branches, etc.) that heavily depend on some pending issue resolution will naturally grow that backlog, especially when the exact resolution of that issue is unknown and may impact all those changes quite a bit!

BTW, creating lots of PRs that fix a duplicated issue in a new/untested way is likely to face very similar problems.

I cannot stop new PRs, but I do not understand why keep creating all that extra work/overhead.

>  please advise if you have any idea on how I can keep myself unblocked otherwise

My recommendation is to work towards spaces() resolution. If you have free time after that, pick a non-YAML problem that does _not_ depend on spaces() resolution.

File: src/event.cc [hunk 2]
```diff
@@ -258,21 +261,28 @@ EventScheduler::clean()
...
+    const auto savedFlags = yaml.flags();
 
+    AtMostOnce header("scheduled events:\n");
     for (auto *e = tasks; e; e = e->next) {
-        out->appendf("%-25s\t%0.3f sec\t%5d\t %s\n",
-                     e->name, (e->when ? e->when - current_dtime : 0), e->weight,
-                     (e->arg && e->cbdata) ? cbdataReferenceValid(e->arg) ? "yes" : "no" : "N/A");
+        yaml << header <<
+             indent << "- operation: " << e->name << '\n' <<
+             indent << indent << "secs to next execution: " << std::setprecision(3) << std::fixed << (e->when? e->when - current_dtime : 0) << '\n';
```
Comment (rousskov) [inline src/event.cc:277]: Please avoid unnecessary abbreviations, especially non-standard ones.

```suggestion
             indent << indent << "seconds to next execution: " << std::setprecision(3) << std::fixed << (e->when ? e->when - current_dtime : 0) << '\n';
```

Many of these events wait for _hours_. I recommend removing "seconds" from the item key so that we can easily change it later, when we improve support for time units:

```suggestion
             indent << indent << "next execution in: " << std::setprecision(3) << std::fixed << (e->when ? e->when - current_dtime : 0) << " seconds\n";
```

... or something like that.

Finally, many of these events do _not_ repeat so "next" is quite misleading here. I would rephrase further to avoid the (often wrong) implication (e.g., "runs in ... seconds" or "scheduled to run in ... seconds").

File: src/event.cc [hunk 3]
```diff
@@ -258,21 +261,28 @@ EventScheduler::clean()
 }
 
 void
-EventScheduler::dump(Packable *out)
+EventScheduler::dump(std::ostream &yaml)
 {
+    static const SBuf indent("  ",2);
```
Comment (rousskov) [inline src/event.cc:266]:
```suggestion
    static const SBuf indent("  ", 2);
```

File: src/event.cc [hunk 4]
```diff
@@ -258,21 +261,28 @@ EventScheduler::clean()
...
+
     if (last_event_ran)
-        out->appendf("Last event to run: %s\n\n", last_event_ran);
+        yaml << "last event to run: " << last_event_ran << "\n";
```
Comment (rousskov) [inline src/event.cc:269]: The current wording is not wrong but is ambiguous. Since we are changing this key, I suggest:

```suggestion
        yaml << "last event that ran: " << last_event_ran << "\n";
```

```suggestion
        yaml << "last event: " << last_event_ran << "\n";
```

... or something like that.

File: src/event.cc (hunk 2 above)
Comment (kinkie) [reply src/event.cc:277]: `... seconds`, from the YAML schema perspective, turns the field from a number into a string, requiring additional work for automations.
In fact, in an ideal world I would specify the unit in a comment for the benefit of humans only: specifying it in the key means that if we change time unit the key changes

Comment (kinkie) [reply src/event.cc:277]: we can use an inline comment:  `next execution in: ... # seconds`

Comment (rousskov) [reply src/event.cc]: > `... seconds`, from the YAML schema perspective, turns the field from a number into a string, requiring additional work for automations.

I strongly disagree with the implication that consuming unitless numbers (or dealing with field key changes) is overall better "for automations" than consuming numbers with units (because the latter requires "additional work"). And since lots of Squid report fields do use units of various kinds, and using units often makes reports more readable for humans, that "additional work" will most likely be required _anyway_ and will be amortized across various consumed reports.

> In fact, in an ideal world I would specify the unit in a comment for the benefit of humans only:

Our ideals differ: Ideally, as a consumer application developer, I would want units to be correctly processed by my code rather than hidden from it in a YAML comment, especially if units change based on the reported value (to assist human readers).

> specifying it in the key means that if we change time unit the key changes

Agreed. The unit should not go into the field key, especially when the unit should change based on the reported value (for human benefit). That is why I explicitly [recommended](https://github.com/squid-cache/squid/pull/1792#discussion_r1583684955) removing PR-added seconds from the field key.
